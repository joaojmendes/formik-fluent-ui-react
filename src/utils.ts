import { FieldProps, getIn } from 'formik'

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export function getErrorMessage<T = any>({ field, form }: FieldProps<T>) {
  const error = getIn(form.errors, field.name)
  const touched = getIn(form.touched, field.name)

  return touched ? error : undefined
}

export function invokeAll(...callbacks: any[]) {
  return () => {
    for (const callback of callbacks) {
      if (callback && typeof callback === 'function') {
        callback()
      }
    }
  }
}

export function createFakeEvent({ name }: { name: string }) {
  return { target: { name } }
}
