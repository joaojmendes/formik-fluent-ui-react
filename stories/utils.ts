import { action } from '@storybook/addon-actions'
import { FormikActions } from 'formik'

export function handleSubmit<T>(
  values: T,
  { setSubmitting }: FormikActions<T>
) {
  action('submit:start')(values)
  setTimeout(() => {
    setSubmitting(false)
    action('submit:done')()
  }, 2000)
}
