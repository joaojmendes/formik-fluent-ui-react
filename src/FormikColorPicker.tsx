import { ColorPicker, IColorPickerProps } from '@fluentui/react';
import { FieldProps } from 'formik'
import * as React from 'react'
import { createFakeEvent, Omit } from './utils'

export function mapFieldToColorPicker<T = any>({
  form,
  field,
}: FieldProps<T>): Pick<IColorPickerProps, 'color' | 'onChange'> {
  return {
    color: field.value,
    onChange: (ev, color) => {
      ev.preventDefault();
      form.setFieldValue(field.name, color),
        field.onBlur(createFakeEvent(field))
    },
  }
}
export type FormikColorPickerProps<T = any> = Omit<
  IColorPickerProps,
  'color' | 'onChange'
> &
  FieldProps<T>

export function FormikColorPicker<T = any>({
  field,
  form,
  ...props
}: FormikColorPickerProps<T>) {
  return (
    <ColorPicker
      {...props}
      {...mapFieldToColorPicker({
        field,
        form,
      })}
    />
  )
}
