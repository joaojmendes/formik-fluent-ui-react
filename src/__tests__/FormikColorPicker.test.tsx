import { Field, FieldProps, Form, Formik } from 'formik'
import { ColorPicker } from 'office-ui-fabric-react'
import * as React from 'react'
import renderer from 'react-test-renderer'
import { FormikColorPicker, mapFieldToColorPicker } from '../FormikColorPicker'
import { noop, serialize } from './utils'

class Values {
  public color: string = '#000000'
}

function createFieldProps(value: string = '#000000'): FieldProps<Values> {
  return {
    field: {
      value,
      onChange: jest.fn(),
      onBlur: jest.fn(),
      name: 'color',
    },
    form: { setFieldValue: jest.fn(), handleBlur: jest.fn(() => jest.fn()) },
  } as any
}

test('<FormikColorPicker /> renders correctly as a field component', () => {
  const component = renderer.create(
    <Formik initialValues={new Values()} onSubmit={noop}>
      <Form>
        <Field name="color" component={FormikColorPicker} />
      </Form>
    </Formik>
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('<FormikColorPicker /> renders a Fabric <ColorPicker />', () => {
  const fieldProps = createFieldProps()

  const formikColorPicker = renderer.create(
    <FormikColorPicker {...fieldProps} />
  )
  const fabricColorPicker = renderer.create(
    <ColorPicker {...mapFieldToColorPicker(fieldProps)} />
  )
  expect(serialize(formikColorPicker)).toBe(serialize(fabricColorPicker))
})

test('mapFieldToColorPicker() maps FieldProps to IColorPickerProps', () => {
  const { field, form } = createFieldProps()
  const props = mapFieldToColorPicker({ form, field })

  expect(props.color).toBe(field.value)



  expect(form.setFieldValue).toHaveBeenCalledTimes(1)
  expect(form.setFieldValue).toHaveBeenCalledWith(field.name, '#ffffff')

  // onColorChanged also calls onBlur
  expect(field.onBlur).toHaveBeenCalled()
})
