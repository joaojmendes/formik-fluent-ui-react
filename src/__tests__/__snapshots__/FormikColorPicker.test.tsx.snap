// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormikColorPicker /> renders correctly as a field component 1`] = `
<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="ms-ColorPicker root-47"
  >
    <div
      className="ms-ColorPicker-panel panel-48"
    >
      <div
        className="ms-ColorPicker-colorRect root-53"
        onMouseDown={[Function]}
        style={
          Object {
            "backgroundColor": "#ff0000",
            "minHeight": 220,
            "minWidth": 220,
          }
        }
      >
        <div
          className="ms-ColorPicker-light light-54"
        />
        <div
          className="ms-ColorPicker-dark dark-55"
        />
        <div
          className="ms-ColorPicker-thumb thumb-56"
          style={
            Object {
              "backgroundColor": "#000000",
              "left": "0%",
              "top": "100%",
            }
          }
        />
      </div>
      <div
        className="ms-ColorPicker-slider is-hue root-57"
        onMouseDown={[Function]}
        style={
          Object {
            "background": "linear-gradient(to left,red 0,#f09 10%,#cd00ff 20%,#3200ff 30%,#06f 40%,#00fffd 50%,#0f6 60%,#35ff00 70%,#cdff00 80%,#f90 90%,red 100%)",
          }
        }
      >
        <div
          className="ms-ColorPicker-sliderOverlay sliderOverlay-58"
        />
        <div
          className="ms-ColorPicker-thumb is-slider sliderThumb-59"
          style={
            Object {
              "left": "0%",
            }
          }
        />
      </div>
      <div
        className="ms-ColorPicker-slider is-alpha root-57"
        onMouseDown={[Function]}
        style={
          Object {
            "backgroundImage": "url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAJUlEQVQYV2N89erVfwY0ICYmxoguxjgUFKI7GsTH5m4M3w1ChQC1/Ca8i2n1WgAAAABJRU5ErkJggg==)",
          }
        }
      >
        <div
          className="ms-ColorPicker-sliderOverlay sliderOverlay-58"
          style={
            Object {
              "background": "linear-gradient(to right, transparent 0, #000000 100%)",
            }
          }
        />
        <div
          className="ms-ColorPicker-thumb is-slider sliderThumb-59"
          style={
            Object {
              "left": "100%",
            }
          }
        />
      </div>
      <table
        cellPadding="0"
        cellSpacing="0"
        className="ms-ColorPicker-table table-49"
      >
        <thead>
          <tr
            className="tableHeader-50"
          >
            <td
              className="tableHexCell-51"
            >
              Hex
            </td>
            <td>
              Red
            </td>
            <td>
              Green
            </td>
            <td>
              Blue
            </td>
            <td>
              Alpha
            </td>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <div
                className="ms-TextField ms-ColorPicker-input input-61"
              >
                <div
                  className="ms-TextField-wrapper wrapper-62"
                >
                  <div
                    className="ms-TextField-fieldGroup fieldGroup-63"
                  >
                    <input
                      aria-invalid={false}
                      aria-label="Hex"
                      className="ms-TextField-field field-64"
                      id="TextField1"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onInput={[Function]}
                      spellCheck={false}
                      type="text"
                      value="000000"
                    />
                  </div>
                </div>
              </div>
            </td>
            <td
              style={
                Object {
                  "width": "18%",
                }
              }
            >
              <div
                className="ms-TextField ms-ColorPicker-input input-61"
              >
                <div
                  className="ms-TextField-wrapper wrapper-62"
                >
                  <div
                    className="ms-TextField-fieldGroup fieldGroup-63"
                  >
                    <input
                      aria-invalid={false}
                      aria-label="Red"
                      className="ms-TextField-field field-64"
                      id="TextField3"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onInput={[Function]}
                      spellCheck={false}
                      type="text"
                      value="0"
                    />
                  </div>
                </div>
              </div>
            </td>
            <td
              style={
                Object {
                  "width": "18%",
                }
              }
            >
              <div
                className="ms-TextField ms-ColorPicker-input input-61"
              >
                <div
                  className="ms-TextField-wrapper wrapper-62"
                >
                  <div
                    className="ms-TextField-fieldGroup fieldGroup-63"
                  >
                    <input
                      aria-invalid={false}
                      aria-label="Green"
                      className="ms-TextField-field field-64"
                      id="TextField5"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onInput={[Function]}
                      spellCheck={false}
                      type="text"
                      value="0"
                    />
                  </div>
                </div>
              </div>
            </td>
            <td
              style={
                Object {
                  "width": "18%",
                }
              }
            >
              <div
                className="ms-TextField ms-ColorPicker-input input-61"
              >
                <div
                  className="ms-TextField-wrapper wrapper-62"
                >
                  <div
                    className="ms-TextField-fieldGroup fieldGroup-63"
                  >
                    <input
                      aria-invalid={false}
                      aria-label="Blue"
                      className="ms-TextField-field field-64"
                      id="TextField7"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onInput={[Function]}
                      spellCheck={false}
                      type="text"
                      value="0"
                    />
                  </div>
                </div>
              </div>
            </td>
            <td
              style={
                Object {
                  "width": "18%",
                }
              }
            >
              <div
                className="ms-TextField ms-ColorPicker-input input-61"
              >
                <div
                  className="ms-TextField-wrapper wrapper-62"
                >
                  <div
                    className="ms-TextField-fieldGroup fieldGroup-63"
                  >
                    <input
                      aria-invalid={false}
                      aria-label="Alpha"
                      className="ms-TextField-field field-64"
                      id="TextField9"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onInput={[Function]}
                      spellCheck={false}
                      type="text"
                      value="100"
                    />
                  </div>
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</form>
`;
